### PR TITLE
LayerNormFusion - Cast support

### PR DESCRIPTION
**Pull request by ytaous**
**Description**: In latest pytorch, Cast nodes may be inserted before Pow to ensure that both inputs 'base' and 'power' are the same type due to restriction in older opsets. The matching pattern in https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/optimizer/layer_norm_fusion.cc needs to be updated to adapt to this additional case.
To repro:

import torch
m = torch.nn.LayerNorm([3, 3])
x = torch.randn(32, 3, 3)
torch.onnx.export(m, (x,), 'layer_norm.onnx')

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: onnxruntime/core/optimizer/layer_norm_fusion.cc [hunk 1]
```diff
@@ -215,33 +254,51 @@ Status LayerNormFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level,
...
     if (scale == nullptr || bias == nullptr) {
       continue;
     }
 
     // Scale and bias must have the same dimension.
-    if (scale->Shape()->dim(0).dim_value() != bias->Shape()->dim(0).dim_value()) {
-      continue;
+    for (int i = 0; i < scale->Shape()->dim_size(); i++) {
+      if (scale->Shape()->dim(i).dim_value() != bias->Shape()->dim(i).dim_value()) {
+        continue;
```
Comment (BowenBao) [inline onnxruntime/core/optimizer/layer_norm_fusion.cc:298]: the `continue` here is not able to proceed onto the outer loop, with the new loop that is wrapped around it.

File: onnxruntime/core/optimizer/layer_norm_fusion.cc [hunk 2]
```diff
@@ -66,6 +76,22 @@ Status LayerNormFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level,
...
+    // verify axes is contiguous if more than one elements
+    const onnxruntime::NodeAttributes& attributes = reduce_mean_node.GetAttributes();
+    std::vector<int64_t> axes_values;
+    if (attributes.find("axes") != attributes.end()) {
+      axes_values = RetrieveValues<int64_t>(attributes.at("axes"));
+      if (axes_values.size() > 1) {
+        sort(axes_values.begin(), axes_values.end());
+        for (size_t i = 0; i < axes_values.size() - 1; i++) {
+          if (axes_values[i] + 1 != axes_values[i + 1]) {
+            continue;
```
Comment (BowenBao) [inline onnxruntime/core/optimizer/layer_norm_fusion.cc:89]: Same as the other comment. the `continue` might not work as intended.

File: onnxruntime/core/optimizer/layer_norm_fusion.cc [hunk 3]
```diff
@@ -66,6 +76,22 @@ Status LayerNormFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level,
         !IsSupportedDataType(reduce_mean_node)) {
       continue;
     }
+
+    // verify axes is contiguous if more than one elements
```
Comment (BowenBao) [inline onnxruntime/core/optimizer/layer_norm_fusion.cc:80]: This looks unrelated to the cast issue tho. Please add test cases for valid and invalid `axes` attributes.

File: onnxruntime/core/optimizer/layer_norm_fusion.cc (hunk 2 above)
Comment (ytaous) [reply onnxruntime/core/optimizer/layer_norm_fusion.cc]: good catch

---
In reply to: [444568745](https://github.com/microsoft/onnxruntime/pull/4320#discussion_r444568745) [](ancestors = 444568745)

File: onnxruntime/core/optimizer/layer_norm_fusion.cc (hunk 3 above)
Comment (ytaous) [reply onnxruntime/core/optimizer/layer_norm_fusion.cc:80]: adding one UT for invalid case, valid case has been covered.

File: onnxruntime/core/optimizer/layer_norm_fusion.cc [hunk 4]
```diff
@@ -185,8 +194,22 @@ Status LayerNormFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level,
...
-    const Node* p_sub2_node = graph_utils::FirstParentByType(pow_node, "Sub");
+    // check if Cast node exists between sub and pow
+    const Node* p_cast_node = graph_utils::FirstParentByType(pow_node, "Cast");
+    if (p_cast_node != nullptr) {
+      Node& cast_node = *graph.GetNode(pow_node.InputNodesBegin()->Index());
+      if (!graph_utils::IsSupportedOptypeVersionAndDomain(cast_node, "Cast", {9}) ||
+          cast_node.GetExecutionProviderType() != cast_node.GetExecutionProviderType() ||
+          !optimizer_utils::CheckOutputEdges(graph, cast_node, 1) ||
+          !IsSupportedDataType(cast_node) ||
+          cast_node.GetInputEdgesCount() == 0) {
```
Comment (yufenglee) [inline onnxruntime/core/optimizer/layer_norm_fusion.cc:205]: >GetInputEdgesCount [](start = 20, length = 18)

do we need it

**Comment by tianleiwu** [comment]
There is CastElimination optimizer. For example, if input is float32, and an additional node Cast to float32 will be removed. This change might not need unless it is casted to different data type.

File: onnxruntime/core/optimizer/layer_norm_fusion.cc [hunk 5]
```diff
@@ -212,36 +235,78 @@ Status LayerNormFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level,
...
           bias = last_add_node.MutableInputDefs()[i];
         }
+#endif
       }
     }
     if (scale == nullptr || bias == nullptr) {
       continue;
     }
 
     // Scale and bias must have the same dimension.
```
Comment (tianleiwu) [inline onnxruntime/core/optimizer/layer_norm_fusion.cc:299]: update comment: have same shape.

**Comment by ytaous** [comment]
interesting, let me check. Thx.

---
In reply to: [649723899](https://github.com/microsoft/onnxruntime/pull/4320#issuecomment-649723899) [](ancestors = 649723899)

**Comment by ytaous** [comment]
will keep it since type could be different

---
In reply to: [649726882](https://github.com/microsoft/onnxruntime/pull/4320#issuecomment-649726882) [](ancestors = 649726882,649723899)

File: onnxruntime/core/optimizer/layer_norm_fusion.cc [hunk 6]
```diff
@@ -212,36 +235,78 @@ Status LayerNormFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level,
...
+    // verify axes is contiguous if more than one elements
+    const onnxruntime::NodeAttributes& attributes = reduce_mean_node.GetAttributes();
+    std::vector<int64_t> axes_values;
+    if (attributes.find("axes") != attributes.end()) {
+      axes_values = RetrieveValues<int64_t>(attributes.at("axes"));
+      if (axes_values.size() > 1) {
+        sort(axes_values.begin(), axes_values.end());
+        bool contingous = true;
+        for (size_t i = 0; i < axes_values.size() - 1; i++) {
+          if (axes_values[i] + 1 != axes_values[i + 1]) {
```
Comment (liuziyue) [inline onnxruntime/core/optimizer/layer_norm_fusion.cc:246]: This might not work in some cases when axis is negative. For example, data dim [a, b, c] with axes [1, -1].

Comment (ytaous) [reply onnxruntime/core/optimizer/layer_norm_fusion.cc:246]: checked with Sherlock, when there are multiple axes, the ints should be contiguous for layernorm/reducemean.
I dump additional cases here to onnx model: 
[https://pytorch.org/docs/master/generated/torch.nn.LayerNorm.html#torch.nn.LayerNorm](https://pytorch.org/docs/master/generated/torch.nn.LayerNorm.html#torch.nn.LayerNorm)
all the axes are contiguous. Therefore I am adding the check. 


---
In reply to: [445769645](https://github.com/microsoft/onnxruntime/pull/4320#discussion_r445769645) [](ancestors = 445769645)

**Comment by ytaous** [comment]
The onnx came straight from pytorch lib (below). Unlike the usual .py, e.g., skip_layer_norm_gen.py.
So practically there's no code as such :-)

import torch
m = torch.nn.LayerNorm([3, 3])
x = torch.randn(32, 3, 3)
torch.onnx.export(m, (x,), 'layer_norm.onnx') 

More examples here to replace x and m for different onnx model:
[https://pytorch.org/docs/master/generated/torch.nn.LayerNorm.html#torch.nn.LayerNorm](https://pytorch.org/docs/master/generated/torch.nn.LayerNorm.html#torch.nn.LayerNorm) 

---
In reply to: [437771474](https://github.com/microsoft/onnxruntime/pull/4320#issuecomment-437771474) [](ancestors = 437771474)

File: onnxruntime/core/optimizer/layer_norm_fusion.cc (hunk 6 above)
Comment (liuziyue) [reply onnxruntime/core/optimizer/layer_norm_fusion.cc:246]: I see. I'm asking because the range of "axes" could be [-r, r-1], r being the rank of data. Do you mean that for "axes" attribute, all values in the list should be either all positive or all negative? In theory reducing with axes [1, -1] on data rank of 3 like my example mentioned above is equivalent to [1, 2] and is also contiguous, but here will return false.

Comment (ytaous) [reply onnxruntime/core/optimizer/layer_norm_fusion.cc:246]: I see your point, let me include this in the check logic. thx.

File: onnxruntime/core/optimizer/layer_norm_fusion.cc [hunk 7]
```diff
@@ -212,36 +234,82 @@ Status LayerNormFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level,
     }
     nodes_to_remove.push_back(last_add_node);
 
+    // verify axes is contiguous if more than one elements
+    const onnxruntime::NodeAttributes& attributes = reduce_mean_node.GetAttributes();
+    std::vector<int64_t> axes_values, normalized_axes;
+    if (attributes.find("axes") != attributes.end()) {
+      axes_values = RetrieveValues<int64_t>(attributes.at("axes"));
+      if (axes_values.size() > 1) {
+        const int input_dim_size = reduce_mean_node.MutableInputDefs()[0]->Shape()->dim_size();
```
Comment (liuziyue) [inline onnxruntime/core/optimizer/layer_norm_fusion.cc:243]: Do we need to check if shape() is nullptr? Shape info is not always available. If no inferred shape, we can just use 0 for input_dim_size.

Comment (ytaous) [reply onnxruntime/core/optimizer/layer_norm_fusion.cc]: on second thought, I'd like to keep it simple. checking axes value is adding unnecessary complexity. So I am dropping the original logic.

---
In reply to: [445825277](https://github.com/microsoft/onnxruntime/pull/4320#discussion_r445825277) [](ancestors = 445825277)
